packages/teuchos/parameterlist/src/Teuchos_YamlParser.cpp

```diff
@@ -1076,47 +1231,279 @@ namespace YAMLParameterList
+  auto yaml_input = ::YAML::LoadAll(yaml);
+  return readParams(yaml_input);
+#else
   any result;
+  Teuchos::YAMLParameterList::Reader reader;
   reader.read_stream(result, yaml, "parseYamlStream");
   ParameterList& pl = any_ref_cast<ParameterList>(result);
   return Teuchos::rcp(new ParameterList(pl));
+#endif // HAVE_TEUCHOSPARAMETERLIST_YAMLCPP
+}
+
+// The following three functions (readParams, processMapNode, and processKeyValueNode)
+// were previously removed from Trilinos in PR 1779 (Teuchos: use Parser, not yaml-cpp, to read YAML PL).
+
+#ifdef HAVE_TEUCHOSPARAMETERLIST_YAMLCPP
+
+Teuchos::RCP<Teuchos::ParameterList> readParams(std::vector<::YAML::Node>& lists)
+{
+  Teuchos::RCP<Teuchos::ParameterList> pl = rcp(new Teuchos::ParameterList); //pl is the root ParameterList to be returned
+  // If there is exactly one element in "lists", assume it is the anonymous top-level parameter list
+  // If there are more than one, place them all in the anonymous top-level list
+  for(size_t i = 0; i < lists.size(); i++)
+  {
+    processMapNode(lists[i], *pl, true);
+  }
+  return pl;
+}
+
+void processMapNode(const ::YAML::Node& node, Teuchos::ParameterList& parent, bool topLevel)
+{
+  if (node.Type() != ::YAML::NodeType::Map)
+  {
+    throw YamlStructureError("All top-level elements of the YAML file must be maps.");
+  }
+  if (topLevel)
+  {
+    parent.setName(node.begin()->first.as<std::string>());
+    processMapNode(node.begin()->second, parent);
+  }
+  else
+  {
+    for (::YAML::const_iterator i = node.begin(); i != node.end(); i++)
+    {
+      // make sure the key type is a string
+      if(i->first.Type() != ::YAML::NodeType::Scalar)
+      {
+        throw YamlKeyError("Keys must be YAML scalars (int, double, or string)");
+      }
+      // if this conversion fails and throws for any reason (shouldn't), let the caller handle it
+      const std::string key = quoted_as<std::string>(i->first);
+      processKeyValueNode(key, i->second, parent, topLevel);
+    }
+  }
+}
+
+void processKeyValueNode(const std::string& key, const ::YAML::Node& node, Teuchos::ParameterList& parent, bool topLevel)
+{
+  // node (value) type can be a map (for nested param lists),
+  // a scalar (int, double, string), or a sequence of doubles (vector<double>)
+  if(node.Type() == ::YAML::NodeType::Scalar)
+  {
+    try
+    {
+      safe_set_entry<int>(parent, key, quoted_as<int>(node));
+    }
+    catch(...)
+    {
+      try
+      {
+        safe_set_entry<double>(parent, key, quoted_as<double>(node));
+      }
+      catch(...)
+      {
+        std::string rawString = quoted_as<std::string>(node);
+        if(rawString == "true")
+        {
+          safe_set_entry<bool>(parent, key, true);
+        }
+        else if(rawString == "false")
+        {
+          safe_set_entry<bool>(parent, key, false);
+        }
+        else
+        {
+          safe_set_entry<std::string>(parent, key, rawString);
+        }
+      }
```

Comment:
There used to be an extra `catch(...)` here that threw a `YamlScalarError` in the event that the provided value was not a int, double, bool, or string. However, the `if` statement at line 1323 guarantees that the given value must be a `::YAML::NodeType::Scalar` (ie an int, double, bool, or string) in order to enter this block, so that final catch statement would never be reached. As a result, I have removed that exception.